elastic/kibana
# [SecuritySolution] Telemetry for data quality dashboard

## Summary

https://github.com/elastic/security-team/issues/6531
Add an extra field to telemetry https://github.com/elastic/kibana/pull/163937





### Success Criteria [Telemetry]
- [x] Date and time of the check
- [x] Number of indices that were checked
- [x] Time it took for each quality check to complete
- [x] Number of failures, if any
- [x] Number of incompatible fields per index
- [x] Identifying the fields that are incompatible to spot recurring mapping conflicts in customers' environments - an array of incompatible fields
- [x] Identifying unallowed values to determine patterns of unallowed values found in customers' environments - an array of incompatible fields
- [x] Number of documents per cluster, index pattern, and index - pattern not tracked, index is tracked by uuid
- [x] Storage information per cluster, index pattern, and index - pattern not tracked, index is tracked by uuid


### Additional properties tracked:
1. batchId: events triggered by the same action share the same batchId
2. ecsVersion



[sample data](https://telemetry-v2-staging.elastic.dev/s/securitysolution/app/discover#/view/b0662610-31ff-11ee-adde-d5df298171dd?_g=(filters:!(),refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(columns:!(event_type,properties.pattern,properties.sizeInBytes,properties.numberOfIndices,properties.numberOfDocuments,properties.numberOfIncompatibleFields,properties.timeConsumedMs),filters:!(),grid:(),hideChart:!f,index:c5dc7cd0-2950-4e51-b428-d0451b1b8d9d,interval:auto,query:(language:kuery,query:'event_type%20:%20%22Data%20Quality%20Checked%22%20'),sort:!(!(timestamp,desc))))


### Prerequisit:
```
telemetry.optIn: true
```

### Steps to verify:
Create an index with invalid fields from dev tools, go to data quality dashboard, click `Check All` button.
You should find `Data Quality Index Checked` in your console verbose when each index is checked.
When everything completed, you should find `Data Quality Check All Completed`

```
PUT auditbeat-custom-index-1

PUT auditbeat-custom-index-1/_mapping
{
  "properties": {
    "@timestamp": {
      "type": "date"
    },
    "event.category": {
      "type": "keyword",
      "ignore_above": 1024
    }
  }
}

POST auditbeat-custom-index-1/_doc
{
  "@timestamp": "2023-02-06T09:41:49.668Z",
  "host": {
    "name": "foo"
  },
  "event": {
    "category": "an_invalid_category"
  },
  "some.field": "this",
  "source": {
    "port": 90210,
    "ip": "10.1.2.3"
  }
}

POST auditbeat-custom-index-1/_doc
{
  "@timestamp": "2023-02-06T09:42:22.123Z",
  "host": {
    "name": "bar"
  },
  "event": {
    "category": "an_invalid_category"
  },
  "some.field": "space",
  "source": {
    "port": 867,
    "ip": "10.9.8.7"
  }
}

POST auditbeat-custom-index-1/_doc
{
  "@timestamp": "2023-02-06T09:43:35.456Z",
  "host": {
    "name": "baz"
  },
  "event": {
    "category": "theory"
  },
  "some.field": "for",
  "source": {
    "port": 5,
    "ip": "10.4.6.6"
  }
}

POST auditbeat-custom-index-1/_doc
{
  "@timestamp": "2023-02-06T09:44:36.700Z",
  "host": {
    "name": "@baz"
  },
  "event": {
    "category": "malware"
  },
  "some.field": "rent",
  "source": {
    "port": 309,
    "ip": "10.1.1.1"
  }
}
```



### event_type:

**Data Quality Index Checked**
```
{
      "sizeInBytes": 89517384,
      "numberOfIndices": 1,
      "numberOfIndicesChecked": 1,
      "ilmPhase": "hot",
      "ecsVersion": "8.6.1",
      "numberOfIncompatibleFields": 0,
      "batchId": "43c35a90-e700-45cf-b9e9-822686e16cff",
      "isCheckAll": false, // This is true when it's triggered by check all button clicked
      "unallowedValueFields": [field1, field2],
      "numberOfDocuments": 172327,
      "indexId": "m_liWv7CRGWOuY2Op3lHZw",
      "errorCount": 0,
      "unallowedMappingFields": [field3],
      "timeConsumedMs": 120
    }
```

**Data Quality Check All Completed:**
```
{
      "isCheckAll": true,
      "sizeInBytes": 94708547,
      "numberOfIndices": 4,
      "numberOfIndicesChecked": 4,
      "ecsVersion": "8.6.1",
      "numberOfDocuments": 178049,
      "numberOfIncompatibleFields": 3,
      "batchId": "bdbb47ba-ffbf-432c-8212-8c691145f0d3",
      "timeConsumedMs": 13786
    },
    "timestamp": "2023-08-08T14:39:37.380Z"
  }
```


### Checklist

Delete any items that are not applicable to this PR.
